### PR TITLE
Bitmap fixes

### PR DESCRIPTION
**Pull request by mamikk**
This fixes various issues with the bitmap code when using bitmap_unset() and bitmap_clear(). There is currently no usage of bitmap_unset() and bitmap_clear() in systemd, but since these functions exists they might be used at a later time and hence they should be correct.

### CODE REVIEW

File: src/basic/bitmap.c [hunk 1]
```diff
@@ -83,10 +84,13 @@ int bitmap_set(Bitmap *b, unsigned n) {
...
         if (offset >= b->n_bitmaps) {
-                if (!GREEDY_REALLOC0(b->bitmaps, b->bitmaps_allocated, offset + 1))
+                new_n_bitmaps = offset + 1;
+                if (!GREEDY_REALLOC(b->bitmaps, b->bitmaps_allocated, new_n_bitmaps))
                         return -ENOMEM;
 
-                b->n_bitmaps = offset + 1;
+                /* we need to do memzero here instead of GREEDY_REALLOC0 because the array does not get cleared on bitmap_clear */
+                memzero(b->bitmaps + b->n_bitmaps, (new_n_bitmaps - b->n_bitmaps) * sizeof(b->bitmaps[0]));
+                b->n_bitmaps = new_n_bitmaps;
```
Comment (zonque) [inline src/basic/bitmap.c:93]: Nice catch! However, the better fix is to really free the memory in `bitmap_clear()` and also set `b->bitmaps_allocated = 0`. Then this function does the right thing as it stands.

File: src/test/test-bitmap.c [hunk 2]
```diff
@@ -101,5 +101,23 @@ int main(int argc, const char *argv[]) {
...
+        bitmap_unset(b, 0);
+        assert_se(bitmap_equal(b, b2));
+
+        assert_se(bitmap_set(b, 1) == 0);
+        bitmap_clear(b);
+        assert_se(bitmap_equal(b, b2));
+
+        assert_se(bitmap_set(b, 0) == 0);
+        assert_se(bitmap_set(b2, 0) == 0);
+        assert_se(bitmap_equal(b, b2));
```
Comment (zonque) [inline src/test/test-bitmap.c:120]: Yay, more tests are much appreciated!

File: src/basic/bitmap.c [hunk 3]
```diff
@@ -191,8 +197,19 @@ bool bitmap_equal(Bitmap *a, Bitmap *b) {
...
-        return memcmp(a->bitmaps, b->bitmaps, sizeof(uint64_t) * a->n_bitmaps) == 0;
+        if (a->n_bitmaps > common_n_bitmaps) {
+                for (i = common_n_bitmaps; i < a->n_bitmaps; i++)
+                        if (a->bitmaps[i] != 0)
+                                return false;
+        } else if (b->n_bitmaps > common_n_bitmaps) {
+                for (i = common_n_bitmaps; i < b->n_bitmaps; i++)
+                        if (b->bitmaps[i] != 0)
+                                return false;
+        }
```
Comment (dvdhrm) [inline src/basic/bitmap.c:212]: Code looks good, but I'd prefer if we make it look a bit simpler. Something like this:

``` c
Bitmap *c;

c = a->n_bitmaps > b->n_bitmaps ? a : b;
for (i = common_n_bitmaps; i < c->n_bitmaps; ++i)
        if (c->bitmaps[i])
                return false;

return true;
```

**Comment by mamikk** [comment]
Whats the preferred way of updating this pull request? Should I do the fixes as additional commits into this pull request or should I create a new pull request using a rebased fixed version of the commits?

**Comment by zonque** [comment]
> Whats the preferred way of updating this pull request? Should I do the fixes as additional commits into this pull request or should I create a new pull request using a rebased fixed version of the commits?

Close this one, and open another one for v2. Thanks!

**Pull request closed without merging**